wgpu-core/src/command/query.rs

```diff
@@ -0,0 +1,441 @@
+    reset_state: &mut Option<&mut QueryResetMap>,
+    invalid_query_set_fn: impl FnOnce(id::QuerySetId) -> E,
+) -> Result<(&'a QuerySet<B>, hal::query::Query<'a, B>), E>
+where
+    E: From<QueryUseError>,
+{
+    let query_set = tracker
+        .use_extend(storage, query_set_id, (), ())
+        .map_err(|e| match e {
+            UseExtendError::InvalidResource => invalid_query_set_fn(query_set_id),
+            _ => unreachable!(),
+        })?;
+
+    // We need to defer our resets because we are in a renderpass, add the usage to the reset map.
+    if let Some(ref mut reset) = reset_state {
+        let used = reset.use_query_set(query_set_id, query_set, query_index);
+        if used {
+            return Err(QueryUseError::UsedTwiceInsideRenderpass { query_index }.into());
+        }
+    }
+
+    let simple_set_type = SimplifiedQueryType::from(query_set.desc.ty);
+    if simple_set_type != query_type {
+        return Err(QueryUseError::IncompatibleType {
+            query_type,
+            set_type: simple_set_type,
+        }
+        .into());
+    }
+
+    if query_index >= query_set.desc.count {
+        return Err(QueryUseError::OutOfBounds {
+            query_index,
+            query_set_size: query_set.desc.count,
+        }
+        .into());
+    }
+
+    let hal_query = hal::query::Query::<B> {
+        pool: &query_set.raw,
+        id: query_index,
+    };
+
+    Ok((query_set, hal_query))
+}
+
+pub(super) fn validate_and_write_timestamp<B: GfxBackend, E>(
+    cmd_buf_raw: &mut B::CommandBuffer,
+    tracker: &mut ResourceTracker<PhantomData<id::QuerySetId>>,
+    storage: &Storage<QuerySet<B>, id::QuerySetId>,
+    query_set_id: id::QuerySetId,
+    query_index: u32,
+    mut reset_state: Option<&mut QueryResetMap>,
+    invalid_query_set_fn: impl FnOnce(id::QuerySetId) -> E,
+) -> Result<(), E>
+where
+    E: From<QueryUseError>,
+{
+    let (query_set, hal_query) = validate_query(
+        tracker,
+        storage,
+        SimplifiedQueryType::Timestamp,
+        query_set_id,
+        query_index,
+        &mut reset_state,
+        invalid_query_set_fn,
+    )?;
+
+    unsafe {
+        // If we don't have a reset state tracker which can defer resets, we must reset now.
+        if reset_state.is_none() {
+            cmd_buf_raw.reset_query_pool(&query_set.raw, query_index..(query_index + 1));
+        }
+        cmd_buf_raw.write_timestamp(hal::pso::PipelineStage::BOTTOM_OF_PIPE, hal_query);
+    }
+
+    Ok(())
+}
+
+pub(super) fn validate_and_begin_pipeline_statistics_query<B: GfxBackend, E>(
+    cmd_buf_raw: &mut B::CommandBuffer,
+    tracker: &mut ResourceTracker<PhantomData<id::QuerySetId>>,
+    storage: &Storage<QuerySet<B>, id::QuerySetId>,
+    query_set_id: id::QuerySetId,
+    query_index: u32,
+    mut reset_state: Option<&mut QueryResetMap>,
+    active_query: &mut Option<(id::QuerySetId, u32)>,
+    invalid_query_set_fn: impl FnOnce(id::QuerySetId) -> E,
+) -> Result<(), E>
```

Comment:
Because QueryUseError doesn't have the invalid query set values, those are always dealt with at a layer above this one with all the other invalid errors.